src/cli/commands/add.js

```diff
@@ -55,23 +56,22 @@ export class Add extends Install {
    */
   getPatternVersion(pattern: string, pkg: Manifest): string {
     const {exact, tilde} = this.flags;
-    const parts = PackageRequest.normalizePattern(pattern);
+    const {hasVersion, range} = PackageRequest.normalizePattern(pattern);
     let version;
+
     if (getExoticResolver(pattern)) {
       // wasn't a name/range tuple so this is just a raw exotic pattern
       version = pattern;
-    } else if (parts.hasVersion && parts.range) {
+    } else if (hasVersion && range && (semver.satisfies(pkg.version, range) || getExoticResolver(range))) {
```

Comment:
Is the `|| getExoticResolver(range)` required? It should be catched by the `getExoticResolver(pattern)`, right (in the sense that if a pattern isn't exotic, the range probably shouldn't)?